Add Pagination to the Changelog

# Task

Currently, all the entries of the changelog are displayed on a single HTML page at https://ocaml.org/changelog.

Since there are a lot of entries, it makes sense to add pagination:
![Screenshot 2023-10-03 at 15-13-43 OCaml Blog](https://github.com/ocaml/ocaml.org/assets/6594573/d9d45efd-19aa-4541-8418-27e3f57b0606)

As an example of how to do pagination (and for possible code sharing), look at how it is done for the Blog:
1. In `handler.ml`, the list of entries is paginated according to a query parameter "p" from the URL: https://github.com/ocaml/ocaml.org/blob/6a36190efe178a6236a13e4cd79adbf9b42e10c6/src/ocamlorg_web/lib/handler.ml#L191-L218
2. In the template, the pagination component is rendered: https://github.com/ocaml/ocaml.org/blob/6a36190efe178a6236a13e4cd79adbf9b42e10c6/src/ocamlorg_frontend/pages/blog.eml#L67


--

What you need to do:

1. modify the function `changelog` in `handler.ml` to paginate the list of changelog entries.
2. render the pagination component in `changelog.eml`

(Related to https://github.com/ocaml/ocaml.org/issues/1125.)

@sabine little confused around the expression it takes 
```
 exec -- dune build --root .
File "src/ocamlorg_web/lib/handler.ml", line 73, characters 13-43:
73 |   Dream.html (Ocamlorg_frontend.changelog )
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type
         tags:string list ->
         ?current_tag:string -> Data.Changelog.t list -> string
       but an expression was expected of type string
make: *** [all] Error 1

```

how can i modify the arguments this ocaml.frontend.changelog takes?

--- a/src/ocamlorg_frontend/components/pagination.eml
+++ b/src/ocamlorg_frontend/components/pagination.eml
@@ -1,40 +1,45 @@
-let paginated_href ~base_url ~page ~query =
-  base_url ^ "?p=" ^ string_of_int page ^ (if query = "" then "" else "&q=" ^ query)
+let rec construct_query_string = function
+  | h :: t -> "&" ^ (fst h) ^ "=" ^ (snd h) ^ construct_query_string t
+  | [] -> ""
 
-let link ~base_url ~page ~current ~query =
+let paginated_href ~base_url ~page ~queries =
+  base_url ^ "?p=" ^ string_of_int page ^ (construct_query_string
+  queries)
+
+let link ~base_url ~page ~current ~queries =
   let active_class = "border-2 border-current text-primary-600" in
   let non_active_class = "border border-body-100 text-lighter hover:bg-gray-50 hover:no-underline" in
-  <a href="<%s (paginated_href ~base_url ~page ~query) %>"
+  <a href="<%s (paginated_href ~base_url ~page ~queries) %>"
     class="w-10 h-10 lg:w-14 lg:h-14 rounded-lg flex items-center justify-center <%s if page = current then active_class else non_active_class %>">
     <%i page %>
   </a>
 
-let left_arrow ~base_url page ~query=
-  <a href="<%s (paginated_href ~base_url ~page:(page - 1) ~query) %>"
+let left_arrow ~base_url page ~queries =
+  <a href="<%s (paginated_href ~base_url ~page:(page - 1) ~queries) %>"
     class="w-10 h-10 lg:w-14 lg:h-14  border border-body-100 rounded-lg flex items-center justify-center text-lighter hover:bg-gray-50 hover:no-underline">
     <%s! Icons.chevron_left "h-5 w-5" %>
   </a>
 
-let right_arrow ~base_url page ~query=
-  <a href="<%s (paginated_href ~base_url ~page:(page + 1) ~query) %>"
+let right_arrow ~base_url page ~queries =
+  <a href="<%s (paginated_href ~base_url ~page:(page + 1) ~queries) %>"
     class="w-10 h-10 lg:w-14 lg:h-14  border border-body-100 rounded-lg flex items-center justify-center text-lighter hover:bg-gray-50 hover:no-underline">
     <%s! Icons.chevron_right "h-5 w-5" %>
   </a>
 
-let render ~total_page_count ~page_number ~base_url ~query=
+let render ~total_page_count ~page_number ~base_url ~queries =
   <div class="pt-16 flex items-center justify-center space-x-2.5">
   <% if  total_page_count > 1 then ( %>
-    <%s! if page_number != 1 then (left_arrow ~base_url page_number ~query) else "" %>
+    <%s! if page_number != 1 then (left_arrow ~base_url page_number ~queries) else "" %>
     <% if page_number >= 4 then ( %>
     <span class="w-10 h-10 lg:w-14 lg:h-14 border-body-100 rounded-lg flex items-center justify-center text-lighter hover:no-underline">...</span>
     <% ); %>
     <% for page = 1 to (total_page_count - 1) do %>
-    <%s! if (abs (page_number - page) <= 1) || (page_number = 1 && page = 3) || (page_number = total_page_count && total_page_count - page = 2) then (link ~base_url ~page ~current:page_number ~query) else "" %>
+    <%s! if (abs (page_number - page) <= 1) || (page_number = 1 && page = 3) || (page_number = total_page_count && total_page_count - page = 2) then (link ~base_url ~page ~current:page_number ~queries) else "" %>
     <% done; %>
     <% if total_page_count - page_number >= 3 then ( %>
     <span class="w-10 h-10 lg:w-14 lg:h-14 border-body-100 rounded-lg flex items-center justify-center text-lighter hover:no-underline">...</span>
     <% ); %>
-    <%s! link ~base_url ~page:total_page_count ~current:page_number ~query %>
-    <%s! if page_number != total_page_count then (right_arrow ~base_url page_number ~query) else "" %>
+    <%s! link ~base_url ~page:total_page_count ~current:page_number ~queries %>
+    <%s! if page_number != total_page_count then (right_arrow ~base_url page_number ~queries) else "" %>
   <% ); %>
   </div>

--- a/src/ocamlorg_frontend/pages/blog.eml
+++ b/src/ocamlorg_frontend/pages/blog.eml
@@ -64,7 +64,7 @@ Layout.render
                     </div>
                     <% ); %>
 
-                    <%s! Pagination.render ~total_page_count:planet_pages_number ~page_number:planet_page ~base_url:Url.blog ~query:""%>
+                    <%s! Pagination.render ~total_page_count:planet_pages_number ~page_number:planet_page ~base_url:Url.blog ~queries:[]%>
                 </div>
             </div>
             <div class="lg:pl-16 space-y-12">

--- a/src/ocamlorg_frontend/pages/changelog.eml
+++ b/src/ocamlorg_frontend/pages/changelog.eml
@@ -1,6 +1,7 @@
-let render ~tags ?current_tag (news : Data.Changelog.t list) =
+let render ~tags ?current_tag ~number_of_pages ~current_page (news : Data.Changelog.t list) =
 Layout.render
 ~title:"OCaml Changelog"
+~canonical:(Url.changelog ^ if current_page = 1 then "" else "?p=" ^ string_of_int current_page)
 ~description:"Read the latest releases and updates from the OCaml ecosystem." @@
 <div class="bg-default dark:bg-dark-default py-24">
   <div class="container-fluid">
@@ -70,6 +71,7 @@ Layout.render
           </div>
         </article>
         <% ); %>
+        <%s! Pagination.render ~total_page_count:number_of_pages ~page_number:current_page ~base_url:Url.changelog ~query:""%>
       </div>
     </div>
     <div class="prose max-w-none mt-8">

--- a/src/ocamlorg_frontend/pages/changelog.eml
+++ b/src/ocamlorg_frontend/pages/changelog.eml
@@ -71,7 +71,7 @@ Layout.render
           </div>
         </article>
         <% ); %>
-        <%s! Pagination.render ~total_page_count:number_of_pages ~page_number:current_page ~base_url:Url.changelog ~query:""%>
+        <%s! Pagination.render ~total_page_count:number_of_pages ~page_number:current_page ~base_url:Url.changelog ~queries:(match current_tag with Some current_tag -> [("t", current_tag)] | None -> [])%>
       </div>
     </div>
     <div class="prose max-w-none mt-8">

--- a/src/ocamlorg_frontend/pages/packages_search.eml
+++ b/src/ocamlorg_frontend/pages/packages_search.eml
@@ -114,7 +114,7 @@ let render ~total ~search ~page ~number_of_pages (packages : Package.package lis
           <% ); %>
         </ol>
         <% ); %>
-        <%s! Pagination.render ~total_page_count:number_of_pages ~page_number:page ~base_url:Url.packages_search ~query:search %>
+        <%s! Pagination.render ~total_page_count:number_of_pages ~page_number:page ~base_url:Url.packages_search ~queries:[("q", search)] %>
       </div>
     </div>
   </div>

--- a/src/ocamlorg_web/lib/handler.ml
+++ b/src/ocamlorg_web/lib/handler.ml
@@ -86,7 +86,7 @@ let changelog req =
           (fun change -> List.exists (( = ) tag) change.Data.Changelog.tags)
           Data.Changelog.all
   in
-  let page, number_of_pages, current_changes = paginate ~req:req ~n:10 changes in
+  let page, number_of_pages, current_changes = paginate ~req:req ~n:30 changes in
   Dream.html (Ocamlorg_frontend.changelog ?current_tag ~tags ~number_of_pages ~current_page:page current_changes)
 
 let changelog_entry req =

--- a/src/ocamlorg_web/lib/handler.ml
+++ b/src/ocamlorg_web/lib/handler.ml
@@ -54,6 +54,22 @@ let community _req =
   let meetups = Data.Meetup.all in
   Dream.html (Ocamlorg_frontend.community ~workshops ~meetups)
 
+let paginate ~req ~n items =
+  let items_per_page = n in
+  let page =
+    Dream.query req "p" |> Option.map int_of_string |> Option.value ~default:1
+  in
+  let number_of_pages =
+    int_of_float
+      (Float.ceil
+         (float_of_int (List.length items) /. float_of_int items_per_page))
+  in
+  let current_items =
+    let skip = items_per_page * (page - 1) in
+    items |> List.drop skip |> List.take items_per_page
+  in
+  (page, number_of_pages, current_items)
+
 let changelog req =
   let current_tag = Dream.query req "t" in
   let tags =
@@ -70,7 +86,8 @@ let changelog req =
           (fun change -> List.exists (( = ) tag) change.Data.Changelog.tags)
           Data.Changelog.all
   in
-  Dream.html (Ocamlorg_frontend.changelog ?current_tag ~tags changes)
+  let page, number_of_pages, current_changes = paginate ~req:req ~n:10 changes in
+  Dream.html (Ocamlorg_frontend.changelog ?current_tag ~tags ~number_of_pages ~current_page:page current_changes)
 
 let changelog_entry req =
   let slug = Dream.param req "id" in
@@ -220,22 +237,6 @@ let workshop req =
     List.find_opt (fun x -> x.Data.Workshop.slug = slug) Data.Workshop.all
   in
   Dream.html (Ocamlorg_frontend.workshop ~videos:watch_ocamlorg_embed workshop)
-
-let paginate ~req ~n items =
-  let items_per_page = n in
-  let page =
-    Dream.query req "p" |> Option.map int_of_string |> Option.value ~default:1
-  in
-  let number_of_pages =
-    int_of_float
-      (Float.ceil
-         (float_of_int (List.length items) /. float_of_int items_per_page))
-  in
-  let current_items =
-    let skip = items_per_page * (page - 1) in
-    items |> List.drop skip |> List.take items_per_page
-  in
-  (page, number_of_pages, current_items)
 
 let blog req =
   let page, number_of_pages, current_items =

--- a/src/ocamlorg_web/lib/handler.ml
+++ b/src/ocamlorg_web/lib/handler.ml
@@ -86,8 +86,10 @@ let changelog req =
           (fun change -> List.exists (( = ) tag) change.Data.Changelog.tags)
           Data.Changelog.all
   in
-  let page, number_of_pages, current_changes = paginate ~req:req ~n:30 changes in
-  Dream.html (Ocamlorg_frontend.changelog ?current_tag ~tags ~number_of_pages ~current_page:page current_changes)
+  let page, number_of_pages, current_changes = paginate ~req ~n:30 changes in
+  Dream.html
+    (Ocamlorg_frontend.changelog ?current_tag ~tags ~number_of_pages
+       ~current_page:page current_changes)
 
 let changelog_entry req =
   let slug = Dream.param req "id" in

